src/ports/postgres/modules/convex/madlib_keras.py_in

```diff
@@ -0,0 +1,633 @@
+    query_result = plpy.execute(model_arch_query)
+    if not  query_result or len(query_result) == 0:
+        plpy.error("no model arch found in table {0} with id {1}".format(model_arch_table, model_arch_id))
+    query_result = query_result[0]
+    model_arch = query_result['model_arch']
+    input_shape = get_input_shape(model_arch)
+    _validate_input_shapes(source_table, independent_varname, input_shape)
+    if validation_table:
+        _validate_input_shapes(
+            validation_table, independent_varname, input_shape)
+    model_weights_serialized = query_result['model_weights']
+
+    # Convert model from json and initialize weights
+    master_model = model_from_json(model_arch)
+    model_weights = master_model.get_weights()
+
+    # Get shape of weights in each layer from model arch
+    model_shapes = []
+    for weight_arr in master_model.get_weights():
+        model_shapes.append(weight_arr.shape)
+
+    if model_weights_serialized:
+        # If warm start from previously trained model, set weights
+        model_weights = deserialize_weights_orig(model_weights_serialized, model_shapes)
+        master_model.set_weights(model_weights)
+
+    end_deserialization = time.time()
+    # plpy.info("Model deserialization time: {} sec".format(end_deserialization - start_deserialization))
+
+    # Construct validation dataset if provided
+    validation_set_provided = bool(validation_table)
+    validation_aggregate_accuracy = []; validation_aggregate_loss = []
+    x_validation = None; y_validation = None
+    if validation_set_provided:
+        x_validation,  y_validation = get_data_as_np_array(validation_table,
+                                                           dependent_varname,
+                                                           independent_varname,
+                                                           input_shape,
+                                                           num_classes)
+
+    # Compute total buffers on each segment
+    total_buffers_per_seg = plpy.execute(
+        """ SELECT gp_segment_id, count(*) AS total_buffers_per_seg
+            FROM {0}
+            GROUP BY gp_segment_id
+        """.format(source_table))
+    seg_nums = [int(each_buffer["gp_segment_id"]) for each_buffer in total_buffers_per_seg]
+    total_buffers_per_seg = [int(each_buffer["total_buffers_per_seg"]) for each_buffer in total_buffers_per_seg]
+
+    # Prepare the SQL for running distributed training via UDA
+    compile_params_to_pass = "$madlib$" + compile_params + "$madlib$"
+    fit_params_to_pass = "$madlib$" + fit_params + "$madlib$"
+    run_training_iteration = plpy.prepare("""
+        SELECT {0}.fit_step(
+            {1}::REAL[],
+            {2}::SMALLINT[],
+            gp_segment_id,
+            {3}::INTEGER,
+            ARRAY{4},
+            ARRAY{5},
+            $MAD${6}$MAD$::TEXT,
```

Comment:
This might be worth investigating in a different JIRA, since we do something similar in other MADlib modules too. So a thorough investigation for a solution we can use elsewhere too might be better.
Created a new JIRA to address this issue:
https://issues.apache.org/jira/browse/MADLIB-1312